### PR TITLE
feat(mesh): gossip CRDT operation log over the wire (d-3a)

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

The mesh-v2 gap analysis (`.claude/docs/mesh/mesh-v2-gap-analysis.md`) flagged the **d-3 cutover** as the single largest gap: CRDT operations never reached the wire. `CrdtNamespace::put`/`delete` only mutated the local op-log, `CrdtOrMap::merge` was reachable only from tests, and the gossip loop carried stream traffic (`td:`/`tree:*`) only. So the `worker:`, `rl:`, and `config:` CRDT namespaces were silent across nodes, and remote CRDT changes never reached local subscribers.

### Solution

Wire the CRDT data path onto the existing `sync_stream` transport (this is **d-3a**; the adapter wiring in `server.rs` and rate-limit enforcement are follow-ups). Two halves, in one PR:

1. **CRDT op-log over the wire** — state converges across nodes.
2. **Subscriber fan-out on remote merge** + value-shape alignment (migration step 7) — remote changes reach subscribers with the same value shape `get` returns.

## Changes

**Wire format** (`crates/mesh/src/proto/gossip.proto`)
- `CrdtOp { key, value, tombstone, timestamp, replica_id }` + `CrdtBatch { ops }` + `CRDT_BATCH` `StreamMessageType` + a `crdt_batch` arm in the `StreamMessage` oneof. `CrdtOp` mirrors the in-crate `Operation`; `replica_id` is the `ReplicaId` UUID in text form. Field-by-field (not an opaque bincode blob) for a language-agnostic schema.

**Codec** (`crates/mesh/src/transport/crdt_batch.rs`, new)
- `build_crdt_batch` (op-log → `CrdtBatch`), `wrap_crdt_batch` (envelope), `dispatch_crdt_batch` (decode → merge). Ops with an unparsable `replica_id` are dropped rather than poisoning the merge.

**Producer / consumer**
- `RoundBatch` gains a `crdt_ops` snapshot, filled in `collect_round_batch` from the store's op-log. Both per-peer senders (`gossip_controller` dialed streams, `gossip_service` accepted streams) broadcast the snapshot each round alongside the stream batch. Merge is idempotent by op-id, so re-sending seen ops is a no-op — **per-peer watermark filtering is a follow-up**.
- Both inbound dispatch blocks gain a `CrdtBatch` arm → `dispatch_crdt_batch`.

**Fan-out + step 7** (`engine/mod.rs`, `engine/lww.rs`, `engine/rate_limit.rs`, `crdt.rs`, `kv.rs`)
- `NamespaceCrdtEngine::apply_remote_ops` now returns `Vec<CrdtChange>`; `CrdtOrMap::merge` concatenates per-engine changes; `MeshKV::merge_crdt_ops` fires `SubscriberRegistry::notify` per change with the canonical post-merge value (matching `get`).
- `CrdtNamespace::put` now notifies the canonical post-insert value instead of the raw caller payload, so local-write and remote-merge subscribers see one shape (the `rl:` adapter no longer needs to special-case the raw payload; its module doc is updated).
- **Change detection emits a `CrdtChange` iff the observable `get(key)` value actually changes** (snapshot before/after the apply loop), not on mere op acceptance. This suppresses spurious events: a newer-version insert rewriting byte-identical bytes, and a tombstone for an already-absent key (`Tombstone→Tombstone` or `vacant→Tombstone`, both encoding to `None`). Generation/log semantics are unchanged.

### Verification

An adversarial multi-agent pass reviewed the change-detection for missed/spurious/wrong-value events. It found **no missed events and no convergence bugs**, and two P2 spurious-event classes — both fixed by the observable-value diff above. (Two findings were correct-as-is: a frontier reshuffle that changes the encoded shard bytes correctly fires, and a benign concurrency TOCTOU where the emitted value is always a valid `get` snapshot and converges.)

## Test Plan

- [x] `cargo test -p smg-mesh --lib` — **165 passed**, 0 failed (+15 over baseline)
  - codec: round-trip, envelope shape, empty-skip, bad-replica-id reject
  - end-to-end (no gRPC, `tests/crdt_integration.rs`): convergence, remote-merge subscriber fan-out, idempotent re-delivery fires nothing, `rl:` canonical-shard shape, tombstone→None
  - `CrdtOrMap::merge` change-detection contract: new value, byte-identical no-op, never-seen tombstone no-op, kill-live emits `None`
- [x] `cargo build -p smg` + `cargo test -p smg --lib mesh::adapters` — pass (downstream unaffected)
- [x] `cargo clippy -p smg-mesh --all-targets` + `cargo fmt` — clean

Net: **+635 / −21** across 18 files.

### Follow-ups (not in this PR)
- Per-peer CRDT send watermark (bandwidth + at-least-once retry).
- Initial CRDT snapshot-on-join / partition heal.
- Wire the v2 adapters (`worker:`/`rl:`/tree) into `server.rs` and connect outbound change paths (d-3 PR 2/3); rate-limit middleware enforcement (d-3b).

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [ ] (Optional) Documentation updated
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

## Summary by CodeRabbit

* **New Features**
  * CRDT operation batches now synchronize across peers via gossip.
  * Subscribers are notified with the canonical post-merge value; notifications are only emitted when a key’s observable value actually changes.
  * Deletes propagate as notifications with a None payload.

* **Tests**
  * Integration tests for remote CRDT delivery, convergence, idempotency, canonical shard shape, and tombstone propagation.
  * Unit tests for change-reporting and tombstone edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/lightseekorg/smg/pull/1570?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: crates/mesh/src/transport/crdt_batch.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+/// Build a [`CrdtBatch`] from an op-log snapshot. Returns `None` for an empty
+/// snapshot so the caller can skip sending an empty message.
+pub fn build_crdt_batch(ops: &[Operation]) -> Option<CrdtBatch> {
+    if ops.is_empty() {
+        return None;
+    }
+    Some(CrdtBatch {
+        ops: ops.iter().map(op_to_proto).collect(),
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/transport/crdt_batch.rs]: Fixed in 2d7ad421. `build_crdt_batches` now splits the op-log snapshot into frames each estimated to stay under `MAX_STREAM_CHUNK_BYTES` (= MAX_MESSAGE_SIZE − 64 KiB envelope margin, the same budget the stream path uses), and both senders loop over the returned batches. A single op larger than the budget is emitted alone (best-effort); worker:/rl:/config: values are far below the cap, so in practice this only bounds the op count per frame.

File: crates/mesh/src/transport/crdt_batch.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    })
+}
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/transport/crdt_batch.rs]: Addressed in 2d7ad421 — went with splitting rather than just a warning: `build_crdt_batches` bounds each frame below MAX_MESSAGE_SIZE (reusing MAX_STREAM_CHUNK_BYTES with its 64 KiB envelope margin), so a large full-op-log broadcast can never produce an oversized frame that tears down the stream. New test `build_crdt_batches_splits_over_budget` covers it.

File: crates/mesh/src/crdt_kv/engine/lww.rs [hunk 3]
```diff
@@ -395,6 +395,19 @@ impl NamespaceCrdtEngine for LwwEngine {
...
+        // Keying off observable value (not op acceptance) suppresses spurious
+        // events from a newer-version op that rewrites byte-identical bytes or
+        // from a dominated op.
+        let mut before: std::collections::HashMap<String, Option<Vec<u8>>> =
+            std::collections::HashMap::new();
+        for op in &unseen {
+            before
+                .entry(op.key().to_string())
+                .or_insert_with(|| self.store.get(op.key()));
+        }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/crdt_kv/engine/lww.rs:411]: Fixed in 2d7ad421 — switched to `if !before.contains_key(op.key()) { before.insert(op.key().to_string(), self.store.get(op.key())) }` so the String is only allocated for keys not yet snapshotted.

File: crates/mesh/src/crdt_kv/engine/rate_limit.rs [hunk 4]
```diff
@@ -377,6 +377,20 @@ impl NamespaceCrdtEngine for RateLimitEngine {
...
+        // that leave `encode_live` unchanged (e.g. Tombstone -> newer
+        // Tombstone, both encoding to `None`) bump generation/state but fire no
+        // subscriber event.
+        let mut before: std::collections::HashMap<String, Option<Vec<u8>>> =
+            std::collections::HashMap::new();
+        for op in &ops {
+            before
+                .entry(op.key().to_string())
+                .or_insert_with(|| self.current_encoded(op.key()));
+        }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/crdt_kv/engine/rate_limit.rs:394]: Fixed in 2d7ad421, same `contains_key` guard as the LWW engine.

File: crates/mesh/src/transport/crdt_batch.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        payload: Some(StreamPayload::CrdtBatch(batch)),
+        sequence,
+        peer_id: self_name.to_owned(),
+    }
+}
+
+/// Receiver-side dispatch for a `CrdtBatch`: decode each op and merge the batch
+/// into the local CRDT store. Ops with an unparsable `replica_id` are skipped.
+/// Merge is idempotent by op-id, so a batch the node has already absorbed is a
+/// no-op. (Subscriber notification on remote merge is a follow-up — d-3a-2.)
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/transport/crdt_batch.rs]: Good catch — fixed in 2d7ad421. The doc now states subscriber fan-out on remote merge is implemented here (via `merge_crdt_ops`), and that an already-absorbed batch is a no-op that fires no event. The d-3a-2 reference was stale leftover from before the fan-out landed in this PR.

File: crates/mesh/src/transport/crdt_batch.rs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+pub fn dispatch_crdt_batch(mesh_kv: &MeshKV, batch: CrdtBatch) {
+    let ops: Vec<Operation> = batch.ops.into_iter().filter_map(proto_to_op).collect();
+    if ops.is_empty() {
+        return;
+    }
+    mesh_kv.merge_crdt_ops(ops);
+}
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/transport/crdt_batch.rs:127]: Deferring this one, with reasoning:

1. **Not CRDT-specific.** `dispatch_stream_batch` (the sibling, already in production) also runs synchronously on the executor and handles larger traffic (tree pages up to 64 MB, chunked). If sync dispatch were a problem it would already be one there. The right fix is a single cross-cutting "non-blocking receive path" change covering both stream and CRDT dispatch, not a CRDT-only bolt-on.

2. **Cost only bites out-of-spec.** The merge is O(batch + log). In the design's supported regime (~30 keys) the op-log is tiny and the merge is microseconds. The "tens of ms" only materializes at the ~10K-op regime, which is already out of the v2 design envelope.

3. **`spawn_blocking` isn't free.** It requires the work to be `Send + 'static` (cloning an `Arc` of state into the closure — a real restructure), pulls a runtime dependency into a currently runtime-agnostic codec layer, and converts a serialized executor path into a concurrent one against the same `CrdtOrMap` (safe given the engines' internal locking, but a semantics change deserving its own review). For the common small-batch case it also *adds* latency.

4. **No measurement justifies it yet** — it's a hypothesis for an out-of-spec regime, and the broadcast-full-op-log behavior is interim (the watermark follow-up shrinks batches to deltas anyway).

Noting it as a candidate for a future cross-cutting non-blocking-dispatch change, gated on an actual profile during the d-3 rollout.
